ast/src/asm_analysis/mod.rs

```diff
@@ -808,28 +784,97 @@ pub struct Rom {
 
 #[derive(Default, Clone, Debug)]
 pub struct AnalysisASMFile {
-    pub items: BTreeMap<AbsoluteSymbolPath, Item>,
+    pub modules: BTreeMap<AbsoluteSymbolPath, Module>,
 }
-
 impl AnalysisASMFile {
-    pub fn machines(&self) -> impl Iterator<Item = (&AbsoluteSymbolPath, &Machine)> {
-        self.items.iter().filter_map(|(n, m)| match m {
-            Item::Machine(m) => Some((n, m)),
-            Item::Expression(_)
-            | Item::TypeDeclaration(_)
-            | Item::TraitDeclaration(_)
-            | Item::TraitImplementation(_) => None,
+    pub fn machines_mut(&mut self) -> impl Iterator<Item = (AbsoluteSymbolPath, &mut Machine)> {
+        self.modules.iter_mut().flat_map(|(module_path, module)| {
+            module.machines.iter_mut().map(move |(name, machine)| {
+                let mut machine_path = module_path.clone();
+                machine_path.push(name.to_string());
+                (machine_path, machine)
+            })
         })
     }
-    pub fn machines_mut(&mut self) -> impl Iterator<Item = (&AbsoluteSymbolPath, &mut Machine)> {
-        self.items.iter_mut().filter_map(|(n, m)| match m {
-            Item::Machine(m) => Some((n, m)),
-            Item::Expression(_)
-            | Item::TypeDeclaration(_)
-            | Item::TraitDeclaration(_)
-            | Item::TraitImplementation(_) => None,
+
+    pub fn machines(&self) -> impl Iterator<Item = (AbsoluteSymbolPath, &Machine)> {
+        self.modules.iter().flat_map(|(module_path, module)| {
+            module.machines.iter().map(move |(name, machine)| {
+                let mut machine_path = module_path.clone();
+                machine_path.push(name.to_string());
+                (machine_path, machine)
+            })
+        })
+    }
+
+    pub fn into_machines(self) -> impl Iterator<Item = (AbsoluteSymbolPath, Machine)> {
+        self.modules.into_iter().flat_map(|(module_path, module)| {
+            module.machines.into_iter().map(move |(name, machine)| {
+                let mut machine_path = module_path.clone();
+                machine_path.push(name.to_string());
+                (machine_path, machine)
+            })
         })
     }
+
+    pub fn get_machine(&self, ty: &AbsoluteSymbolPath) -> Option<&Machine> {
+        let mut path = ty.clone();
```

Comment:
I the clone needed here?

Reply:
Yes since modules are indexed by path and machines by name. To avoid this clone, we'd need slices of `AbsoluteSymbolPath` which is not implemented in this PR.